### PR TITLE
Platform mappings: support setting Starlark flags

### PR DESCRIPTION
**Pull request by gregestren**
In other words, support:

```
platforms:
  //:my_platform
    --//mypkg:my_custom_flag=value
```

#### Non-goals:
 - Refactor the relationship between the Starlark and native option parsers
 - Support Starlark flag -> platform mappings (we could add this if needed but platform mappings isn't a feature we want to expand more than necessary)
 - Parse with simplified ad hoc logic instead of directly using `StarlarkOptionsParser` as the source of truth

#### Approach:
1. The basic problem (described [here](https://github.com/bazelbuild/bazel/issues/18199#issuecomment-1520620249)) is `StarlarkOptionsParser` needs to convert flag names to `Target`s. `PlatformMappingValue` calls the parsers but can't Skyframe-load `Target`s.
1. So we add support in `PlatformMappingFunction` to Skyframe-load the needed `Target`s
1. But `StarlarkOptionParser` already has code for doing that, outside of a `Skyfunction` environment, through `SkyframeExecutor.loadTargetPatternsWithoutFilters`.
1. So factor out `StarlarkOptionParser`s `Target` loading into an interface the caller can implement
1. The original callers (`BlazeOptionHandler` and `CanonicalizeCommand`) provide the same implementation as before
1. `PlatformMappingFunction` provides a Skyframe retriever through `PackageFunction` (which is a cleaner way to get `Target`s within a `Skyfunction`)

I also had to factor out `StarlarkOptionsParser.java` into its own `java_library` to prevent dependency cycles between `:runtime` and Skyframe: `:runtime` already depends on Skyframe, and by calling `StarlarkOptionsParser` from a Skyfunction that adds a dependency the other way.

Thanks @katre for initial investigation.

Fixes https://github.com/bazelbuild/bazel/issues/18199.

PiperOrigin-RevId: 528888600
Change-Id: Ie3597f4a1d406c6918fadbdcdb208d23db6a2aa2

### CODE REVIEW

**Comment by gregestren** [comment]
https://github.com/Colecf @jin

**Comment by gregestren** [comment]
I only included a basic test in this first commit. Happy to add more (e.g. bad label checking).

File: src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java [hunk 1]
```diff
@@ -257,11 +265,88 @@ private void parseArgsAndConfigs(List<String> args, ExtendedEventHandler eventHa
...
+   * <p>This is important because downstream services (like a continuous integration tool or build
+   * results dashboard) read these messages to reconcile which requested targets were built. If they
+   * determine Blaze tried to build {@code //foo //bar} then see a "target configured" message for
+   * some other target {@code //my_starlark_flag}, they might show misleading messages like "Built 3
+   * of 2 requested targets.".
+   *
+   * <p>Hence this class. By dropping those events, we restrict all info and error reporting logic
+   * to the options parsing pipeline.
+   */
+  private static class NonPostingEventHandler implements ExtendedEventHandler {
```
Comment (gregestren) [inline src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java:286]: Note this is just ported out of `StarlarkOptionsParser`.

File: src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java [hunk 2]
```diff
@@ -257,11 +265,88 @@ private void parseArgsAndConfigs(List<String> args, ExtendedEventHandler eventHa
...
+        ExtendedEventHandler reporter) {
+      this.skyframeExecutor = skyframeExecutor;
+      this.relativeWorkingDirectory = relativeWorkingDirectory;
+      this.reporter = new NonPostingEventHandler(reporter);
+    }
+
+    @Override
+    public Target loadBuildSetting(String targetLabel)
+        throws InterruptedException, TargetParsingException {
+      TargetPatternPhaseValue result =
```
Comment (gregestren) [inline src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java:333]: This implementation code is ported out of `StarlarkOptionsParser`.

**Comment by jin** [comment]
Thanks @gregestren! Do you mind adding a more comprehensive test that changes the platform in a transition, and verifies that the associated build flag with the platform also changes value?

File: src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java [hunk 3]
```diff
@@ -257,11 +265,88 @@ private void parseArgsAndConfigs(List<String> args, ExtendedEventHandler eventHa
...
+    public void post(ExtendedEventHandler.Postable e) {}
+  }
+
+  /**
+   * Lets {@link StarlarkOptionsParser} convert flag names to {@link Target}s through {@link
+   * TargetPatternPhaseValue}.
+   *
+   * <p>This is used for top-level flag parsing, outside any {@link SkyFunction}.
+   */
+  public static class SkyframeExecutorTargetLoader
```
Comment (katre) [inline src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java:308]: Do we need a new inner class? Can the BlazeOptionHandler implement BuildSettingLoader directly, or just define `loadBuildSetting` in BlazeOptionHandler and pass in a method reference?

**Review by katre — APPROVED** [review]
Overall LGTM.

I would have preferred this split into a few PRs:
1. Move StarlarkOptionsParser to a new target and introduce the BuildSettingsLoader interface (and existing usage).
  1. This would have introduced no new behavior, making testing simpler.
3. Add new usage of StarlarkOptions in platform mapping.
  1. New behavior, so new tests.

But this is written so it's fine as-is.

**Comment by gregestren** [comment]
> Overall LGTM.
> 
> I would have preferred this split into a few PRs:

I was thinking something similar but wanted to at least start with the full picture to see how it all fits together.

This was on the verge of being complicated enough to merit a design doc. But IMO not quite that complicated (since it's not user-facing aside from one very specific difference) and we really don't want to publicize platform mapping upgrades any more than necessary.

File: src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java (hunk 3 above)
Comment (gregestren) [reply src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java:308]: `CanonicalizeCommand` also references this, which is why I made it its own entity: https://github.com/bazelbuild/bazel/blob/f2ccd3c1362546b8d346ed19a1c6eff3951dfce9/src/main/java/com/google/devtools/build/lib/runtime/commands/CanonicalizeCommand.java#L201

**Comment by gregestren** [comment]
> Thanks @gregestren! Do you mind adding a more comprehensive test that changes the platform in a transition, and verifies that the associated build flag with the platform also changes value?

On it...

Also, I'm not sure how this handles labels outside the main repo. I did my best to use canonical label references but all that logic is so subtle and you never know.

This is complicated enough where I'd like to not worry about that for this PR. If you think you'd need that let me know and let's explore as a followup patch (or even better if no patch is needed at all).

**Comment by gregestren** [comment]
@brentleyjones out of curiosity, what's your interest in this change?

**Comment by brentleyjones** [comment]
Not out if need (that I know of).

I'm just a fan of Starlark build settings/flags, so I like changes like this that make them work in more places.

**Comment by gregestren** [comment]
> > Thanks @gregestren! Do you mind adding a more comprehensive test that changes the platform in a transition, and verifies that the associated build flag with the platform also changes value?
> 

Added tests.

**Review by jin — APPROVED** [review]
Thank you for the test case! @Colecf FYI

I'll leave the review of the impl details to John.

**Comment by katre** [comment]
LGTM for me.

**Comment by Colecf** [comment]
Thanks Greg! I'll try and pull this into android and test it out.

**Pull request closed without merging**